File: problems/1/scala/1.scala
import scala.math._

object HelloWorld {

    // Simple solution using modulo arithmetic
    def sumMultipleModulo(number: Int, factors: List[Int]): Int = {
        var sum = 0

        for (n <- 1 to number) {
            if (factors exists { n % _ == 0}){
                sum = sum + n
            }
        }
        sum
    }

    // Simple solution using modulo arithmetic
    def sumMultipleFormula(number: Int, factors: List[Int]): Int = {
        val sumFactor = (f:Int):Int => {
            val s = loor(number / f)
            val l = f + ((s - 1) * f)
            val sum = s * (f + l) / 2
        }
        sumFactor(1)
    }

    // Using formulae for arithmetic progressions
    def main(args: Array[String]): Unit = {
        val sum = sumMultipleModulo(10, List(3,5))
        println(sum)
    }
}

File: problems/1/js/1.js
const benchmark = require('../../../utils/js/benchmark')

const n = 1000;

// Simple solution using modulo arithmetic
const sumMultipleModulo = ({number, factors}) => {
    let sum = 0;
    for (let i = 0; i <= number; i++) {
        if (factors.some(f => i % f === 0)) {
            sum += i; 
        }
    }
    return sum;
}

// Using formulae for arithmetic progressions
const sumMultipleFormula = ({number, factors}) => {
    const sumFactor = f => {
        let s = Math.floor(number / f);
        let l = f + ((s - 1) * f);
        return s * (f + l) / 2
    }
    
    return factors
        .map(sumFactor)
        .reduce((acc, curr) => acc += curr, 0) - sumFactor(
            factors.reduce((acc, curr) => acc * curr, 1)
        )
}

benchmark([
    () => sumMultipleModulo({
        number: n,
        factors: [3, 5]
    }),
    () => sumMultipleFormula({
        number: n,
        factors: [3, 5]
    })
]);


